### DISCUSSION TITLE
Option to Turn off Designer Code Formatting

### DISCUSSION THREAD

**Discussion by AssisiSoft** [dotnet/winforms#8754] [Ideas]
My coding format uses tabs for aligning various kinds of statements like assignments, breaks, error propagation and more.  There are no formatting settings that achieve this kind of format so I have to resort to turning off all formatting that I can.  Unfortunately both Winforms and Dataset designers do not have an option to turn of auto formatting.  Now, in VS 17.5, even "code behind" files get reformatted.  Immediately after a form is opened in the designer!  This is a total stop for me so I am moving back to 17.4.  There is over a million lines of code that I have written that need to not be re-formatted: I have tabbed and organized it as I need it to be.

Can we **please** have a setting for turning off auto formatting?  I would like to turn it off everywhere if possible.  It seems like it should be a simple options to implement.

**Comment by RussKie** [comment]
The *.Designer.cs file is auto-generated, and it has a note at the top of the file to that effect.

**Reply by AssisiSoft** [reply]
Talking about "partial" files not the designer files.  There was a time in the past when VS did not reformat user code in partial class files.

We need a global option for turning off auto formatting for all files that are **user** files.  Including "partial class files" that are associated with designers but are not created by designers.

**Comment by dreddy-work** [comment]
@AssisiSoft, what do you mean by partial classes associated with designer but are not auto generated? Can you attach a simple application here for reference?

**Comment by AssisiSoft** [comment]
Below is a class EditorMap, defined in two files: my code in EditorMap.cs and Winforms designer code in EditorMap.designer.cs.  In versions of VS prior to 17.5, the winforms designer would reformat its designer code in “EditorMap.designer.cs” but leave my code in “EditorMap.cs” alone.  Version 17.5 reformats both files now.  We need an option to turn off auto format at least for our edited code.  Preferably for all files if desired.


EditorMap.cs…
internal partial class EditorMap : Editor
{
//My Code in my own preferred format please!
//Do not autoformat this code.  The Designer is not writing it, I am.
}

EditorMap.designer.cs…
internal partial class EditorMap : Editor
{
//Designer code in designer’s format.
}


Rich Howard
Assisi Software
503.233.6417
***@***.******@***.***>
www.AssisiSoft.com<http://www.assisisoft.com/>
***@***.***
Online Resources
User Forum<http://www.assisisoft.com/forum>
Library<http://www.assisisoft.com/library>
Version History<http://www.assisisoft.com/versionhistory>

**Comment by KlausLoeffelmann** [comment, marked as answer]
Yes, that's a known issue and should be on our internal Designer to do list.

@Olina-Zhang: Can you create a quick repo, where you add code to Form1.cs which is "custom formatted", and we show, that the code gets formatted after writing it. @Tanya-Solyanik, @merriemcgaw, FYI.

As an (internal) suggestion: We could opt out of formatting for C# for code-files which don't contain `InitializeComponent`, because we never touch them to begin with. We should for the time being continue doing this for VB, since we're touching the code when we're updating the event handler method stubs.

**Comment by Olina-Zhang** [comment]
@KlausLoeffelmann It is a similar issue in our internal designer repo: https://github.com/microsoft/winforms-designer/issues/5079

**Reply by simmotech** [reply]
That url doesn't exist.

**Reply by KlausLoeffelmann** [reply]
It's a link to a Microsoft-internal repo, for the team's information.
Thanks, @Olina - we have this on our radar!

**Comment by simmotech** [comment]
This isn't just a minor blip. It is a huge f**ing problem!!!
The Designer should not be writing to any file other than the xxx.Designer.cs file. Never Ever.
Give whoever introduced this problem a good kicking from me.

I use Resharper for all my formatting and I thought I was going mad reformatting the same bit of code. I then had the Form side by side with the code, made a trivial change on the Form and changed it back and clicked Save in the toolbar.
Then I actually saw my Resharper formatting disappear.

I hope the OP is right and going to v17.4 will get around the problem.

**Comment by KlausLoeffelmann** [comment]
We will be addressing this in 17.7 and are discussing servicing it back for 17.6.
@merriemcgaw FYI.

**Comment by KlausLoeffelmann** [comment]
> The Designer should not be writing to any file other than the xxx.Designer.cs file. Never Ever.

The Designer needs to write to other files in these cases:
* There is no code-behind file - `InitializeComponent` is part of the main form's class file.
* In the Visual Basic case, because the `Handels` clauses of event procs need to be updated in certain scenarios.

**Reply by simmotech** [reply]
(Reverting to 17.4 does get around the problem for now for anyone interested)

You are right - I meant _formatting_ any other file other than the xxx.Designer.cs file.
But you yourself said higher up "As an (internal) suggestion: We could opt out of formatting for C# for code-files which don't contain InitializeComponent, because we never touch them to begin with"
Well double-clicking on an event creates a method in the non-Designer file I suppose though I don't do that and prefer to add handlers in code.

Actually, I didn't mention a 'code-behind file' but the xxx.Designer.cs _is_ the code-behind file to my mind and according to this:
[Modernization of code-behind in WinForms OOP designer](https://github.com/dotnet/winforms/blob/main/docs/designer/modernization-of-code-behind-in-OOP-designer/modernization-of-code-behind-in-oop-designer.md)

This page also gives a reason - changing to Roslyn for design-time serialization - for the change and it got me thinking:

One of the projects I have to maintain is very old and has literally hundreds of old Forms which don't have a .Designer.cs file (.net 1.0?) - each is in one file. No one wants to go through and manually split them all into two files!
a) Will the new Roslyn serializer still work with that format?
b) It would be a big win if the new Roslyn design-time serializer insisted that the InitializeComponent() goes into a code-behind .Designer.cs class and we could just open and save the file and have it split into two and with the new format in one go.

**Comment by Tanya-Solyanik** [comment]
>old Forms which don't have a .Designer.cs file

The new serialization logic works with forms where InitializeComponent is written to Form1.cs file.

> It would be a big win if the new Roslyn design-time serializer insisted that the InitializeComponent() goes into a code-behind .Designer.cs class and we could just open and save the file and have it split into two and with the new format in one go.

This is a great suggestion. I think we could create an analyzer that splits out InitializeComponent into the .Designer file. @merriemcgaw  - how do we handle suggestions like this?

**Reply by merriemcgaw** [reply]
Interesting idea! I think we can do something to support that.

**Comment by KlausLoeffelmann** [comment]
I agree - we should have an Analyzer and a Code Fixer for this!

To clarify: The change I am currently working on (amongst other things, btw) will limit formatting only to the spots which have actually changed. But everything inside of `InitializeComponent` will _always_ change, no matter where that code is placed. After my PR though, the CodeDOM serializer will not touch anything else that it has not written/modified itself (although we maybe reducing the backing field definition of forms to unify the formatting there).

**Comment by KlausLoeffelmann** [comment]
Unfortunately, the changes where too extensive to integrate into 17.7 that late in the release cycle. The change made it to 17.8 Preview 1, though. We will continue to watch for customer feedback, and may decide at a later date, if we want to service 17.6 with the change, since 17.6 is LTS.

**Comment by AssisiSoft** [comment]
Woo Hoo!

Unfortunately, the changes where too extensive to integrate into 17.7 that late in the release cycle. The change made it to 17.8 Preview 1, though. We will continue to watch for customer feedback, and may decide at a later date, if we want to service 17.6 with the change, since 17.6 is LTS.

**Reply by AssisiSoft** [reply]
Using 17.8 Preview 4 I can verify code in non designer **Form** files are no longer formatted.  Great!  Thank you!
But, partial class files for DataSets **is** reformatted.  Partial class files are **not** designer files and so should not be getting auto formatted.  Can this be added to 17.8 too?
Rich

**Reply by AssisiSoft** [reply]
However...  When an event handler is added via the Properties window, the event is added to the non designer code (my code) and all the code in that file gets reformatted.  Perhaps always write the default event handlers to the designer files?

**Comment by KlausLoeffelmann** [comment]
> But, partial class files for DataSets is reformatted. Partial class files are not designer files and so should not be getting auto formatted. Can this be added to 17.8 too?

Can you please file a bug for this, since it's a different issue?
Please also state, if this is Framework Out-Of-Proc or .NET Out-Of-Proc, and when the reformatting exactly happens.
